Add unsupported types to schema type parser

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28468 Delete ATenDispatch
* #28186 Remove c10->ATen registration forwarding
* #27978 Migrate remaining ops to the c10 dispatcher
* #28208 Move complex extension test to c10
* #28184 Fix codegen for out operators
* #28183 Allow migrating factory methods to c10
* #28182 Fix overload names
* **#28181 Add unsupported types to schema type parser**

These types are needed to parse the schemas from native_functions.yaml.

Note: This doesn't actually add the functionality to JIT, it only makes the parser pass.

Differential Revision: [D17969014](https://our.internmc.facebook.com/intern/diff/D17969014/)

smessmer committed:
Add unsupported types to schema type parser

These types are needed to parse the schemas from native_functions.yaml.

Note: This doesn't actually add the functionality to JIT, it only makes the parser pass.

Differential Revision: [D17969014](https://our.internmc.facebook.com/intern/diff/D17969014/)

[ghstack-poisoned]

smessmer committed:
Update on "Add unsupported types to schema type parser"


These types are needed to parse the schemas from native_functions.yaml.

Note: This doesn't actually add the functionality to JIT, it only makes the parser pass.

Differential Revision: [D17969014](https://our.internmc.facebook.com/intern/diff/D17969014/)

[ghstack-poisoned]

ezyang commented:
I'm pretty sure I don't want ConstQuantizerPtr permanently in the list. @jerryzh168 you have previously told me this name is temporary. What is the plan?

At the VERY least the types we aren't going to stick to should be prominently documented as such.

smessmer committed:
Update on "Add unsupported types to schema type parser"


These types are needed to parse the schemas from native_functions.yaml.

Note: This doesn't actually add the functionality to JIT, it only makes the parser pass.

Differential Revision: [D17969014](https://our.internmc.facebook.com/intern/diff/D17969014/)

[ghstack-poisoned]

smessmer commented:
I added a bunch of TODOs to these types

smessmer committed:
Update on "Add unsupported types to schema type parser"


These types are needed to parse the schemas from native_functions.yaml.

Note: This doesn't actually add the functionality to JIT, it only makes the parser pass.

Differential Revision: [D17969014](https://our.internmc.facebook.com/intern/diff/D17969014/)

[ghstack-poisoned]

smessmer committed:
Update on "Add unsupported types to schema type parser"


These types are needed to parse the schemas from native_functions.yaml.

Note: This doesn't actually add the functionality to JIT, it only makes the parser pass.

Differential Revision: [D17969014](https://our.internmc.facebook.com/intern/diff/D17969014/)

[ghstack-poisoned]

zdevito commented on torch/csrc/jit/script/schema_type_parser.cpp:
```diff
@@ -33,10 +33,14 @@ namespace script {
   static std::unordered_map<std::string, TypePtr> type_map = {
       {"Generator", GeneratorType::get()},
+      {"Dimname", StringType::get()},
+      {"DimnameList", ListType::create(StringType::get())}, // TODO Remove this type from the schema parser and write these as Dimname[] instead
```
This one should not be here. It should be written as`Dimname[]`. We do not have aliases for any other types since it makes the entire type system seem more complicated than it is. We should change it to Dimname[] where they are used and teach the other pathways about it.

smessmer replied:
fixed

zdevito reviewed (commented):
Looks good except for DimnameList.

jerryzh168 commented:
> I'm pretty sure I don't want ConstQuantizerPtr permanently in the list. @jerryzh168 you have previously told me this name is temporary. What is the plan?
> 
> At the VERY least the types we aren't going to stick to should be prominently documented as such.

we'll have a Quantizer type that will replace this one in the future, and it's blocked on TorchBind supporting custom times right now, @jamesr66a has signed up to work on it.

dzhulgakov commented on torch/csrc/jit/script/schema_type_parser.cpp:
```diff
@@ -33,10 +33,14 @@ namespace script {
       {"MemoryFormat", IntType::get()},
-      {"QScheme", IntType::get()},
+      {"Storage", IntType::get()},
+      {"QScheme", IntType::get()},  // TODO This type should be removed from the schema parser, it should use the custom class mechanism instead. @jerryzh
```
this is actually an Enum, not class. So I guess it's similar to MemoryFormat and we can remove this TODO (unless we want to add enums to our typesystem)

dzhulgakov commented on torch/csrc/jit/script/schema_type_parser.cpp:
```diff
@@ -33,10 +33,14 @@ namespace script {
-      {"QScheme", IntType::get()},
+      {"Storage", IntType::get()},
+      {"QScheme", IntType::get()},  // TODO This type should be removed from the schema parser, it should use the custom class mechanism instead. @jerryzh
+      {"ConstQuantizerPtr", IntType::get()},  // TODO This type should be removed from the schema parser, it should use the custom class mechanism instead. @jerryzh
```
yep, I'm fine landing it as is right now though

smessmer committed:
Update on "Add unsupported types to schema type parser"


These types are needed to parse the schemas from native_functions.yaml.

Note: This doesn't actually add the functionality to JIT, it only makes the parser pass.

Differential Revision: [D17969014](https://our.internmc.facebook.com/intern/diff/D17969014/)

[ghstack-poisoned]

smessmer committed:
Update on "Add unsupported types to schema type parser"


These types are needed to parse the schemas from native_functions.yaml.

Note: This doesn't actually add the functionality to JIT, it only makes the parser pass.

Differential Revision: [D17969014](https://our.internmc.facebook.com/intern/diff/D17969014/)

[ghstack-poisoned]

smessmer committed:
Update on "Add unsupported types to schema type parser"


These types are needed to parse the schemas from native_functions.yaml.

Note: This doesn't actually add the functionality to JIT, it only makes the parser pass.

Differential Revision: [D17969014](https://our.internmc.facebook.com/intern/diff/D17969014/)

[ghstack-poisoned]

smessmer committed:
Update on "Add unsupported types to schema type parser"


These types are needed to parse the schemas from native_functions.yaml.

Note: This doesn't actually add the functionality to JIT, it only makes the parser pass.

Differential Revision: [D17969014](https://our.internmc.facebook.com/intern/diff/D17969014/)

[ghstack-poisoned]

smessmer committed:
Update on "Add unsupported types to schema type parser"


These types are needed to parse the schemas from native_functions.yaml.

Note: This doesn't actually add the functionality to JIT, it only makes the parser pass.

Differential Revision: [D17969014](https://our.internmc.facebook.com/intern/diff/D17969014/)

[ghstack-poisoned]

smessmer committed:
Update on "Add unsupported types to schema type parser"


These types are needed to parse the schemas from native_functions.yaml.

Note: This doesn't actually add the functionality to JIT, it only makes the parser pass.

Differential Revision: [D17969014](https://our.internmc.facebook.com/intern/diff/D17969014/)

[ghstack-poisoned]
